File: sso-glide/src/main/java/it/niedermann/nextcloud/sso/glide/SingleSignOnUrl.java
package it.niedermann.nextcloud.sso.glide;

import android.content.Context;

import androidx.annotation.NonNull;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.model.GlideUrl;
import com.bumptech.glide.load.model.Headers;
import com.nextcloud.android.sso.helper.SingleAccountHelper;
import com.nextcloud.android.sso.model.SingleSignOnAccount;

import java.net.URL;

/**
 * Use this as kind of {@link GlideUrl} if you want to do a {@link Glide} request from a {@link SingleSignOnAccount} which is not set by {@link SingleAccountHelper#setCurrentAccount(Context, String)}.
 */
public class SingleSignOnUrl extends GlideUrl {

    public SingleSignOnUrl(@NonNull SingleSignOnAccount ssoAccount, @NonNull String url) {
        this(ssoAccount.name, url);
    }

    public SingleSignOnUrl(@NonNull SingleSignOnAccount ssoAccount, @NonNull URL url) {
        this(ssoAccount.name, url);
    }

    public SingleSignOnUrl(@NonNull SingleSignOnAccount ssoAccount, @NonNull String url, @NonNull Headers headers) {
        this(ssoAccount.name, url, headers);
    }

    public SingleSignOnUrl(@NonNull SingleSignOnAccount ssoAccount, @NonNull URL url, @NonNull Headers headers) {
        this(ssoAccount.name, url, headers);
    }

    public SingleSignOnUrl(@NonNull String ssoAccountName, @NonNull String url) {
        super(url, new SingleSignOnOriginHeader(ssoAccountName));
    }

    public SingleSignOnUrl(@NonNull String ssoAccountName, @NonNull URL url) {
        super(url, new SingleSignOnOriginHeader(ssoAccountName));
    }

    public SingleSignOnUrl(@NonNull String ssoAccountName, @NonNull String url, @NonNull Headers headers) {
        super(url, new SingleSignOnOriginHeader(ssoAccountName, headers));
    }

    public SingleSignOnUrl(@NonNull String ssoAccountName, @NonNull URL url, @NonNull Headers headers) {
        super(url, new SingleSignOnOriginHeader(ssoAccountName, headers));
    }
}

File: sso-glide/src/main/java/it/niedermann/nextcloud/sso/glide/SingleSignOnUrl.kt
package it.niedermann.nextcloud.sso.glide

import android.net.Uri
import com.bumptech.glide.Glide
import com.bumptech.glide.load.model.GlideUrl
import com.bumptech.glide.load.model.Headers
import com.nextcloud.android.sso.helper.SingleAccountHelper
import com.nextcloud.android.sso.model.SingleSignOnAccount
import java.net.URL

/**
 * Use this as kind of [GlideUrl] if you want to do a [Glide] request from a [SingleSignOnAccount] which is not set by [SingleAccountHelper.commitCurrentAccount].
 */
class SingleSignOnUrl : GlideUrl {

    val ssoAccountName: String

    constructor(ssoAccount: SingleSignOnAccount, url: String) : this(ssoAccount.name, url)
    constructor(ssoAccount: SingleSignOnAccount, url: URL) : this(ssoAccount.name, url)
    constructor(ssoAccount: SingleSignOnAccount, uri: Uri) : this(ssoAccount.name, uri.toString())
    constructor(ssoAccount: SingleSignOnAccount, url: String, headers: Headers) : this(ssoAccount.name, url, headers)
    constructor(ssoAccount: SingleSignOnAccount, url: URL, headers: Headers) : this(ssoAccount.name, url, headers)
    constructor(ssoAccount: SingleSignOnAccount, uri: Uri, headers: Headers) : this(ssoAccount.name, uri.toString(), headers)

    constructor(ssoAccountName: String, url: String) : super(url) {
        this.ssoAccountName = ssoAccountName
    }

    constructor(ssoAccountName: String, url: URL) : super(url) {
        this.ssoAccountName = ssoAccountName
    }

    constructor(ssoAccountName: String, uri: Uri) : super(uri.toString()) {
        this.ssoAccountName = ssoAccountName
    }

    constructor(ssoAccountName: String, url: String, headers: Headers) : super(url, headers) {
        this.ssoAccountName = ssoAccountName
    }

    constructor(ssoAccountName: String, url: URL, headers: Headers) : super(url, headers) {
        this.ssoAccountName = ssoAccountName
    }

    constructor(ssoAccountName: String, uri: Uri, headers: Headers) : super(uri.toString(), headers) {
        this.ssoAccountName = ssoAccountName
    }
}